engine/access/rest/websockets/controller_test.go

```diff
@@ -0,0 +1,281 @@
+package websockets
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/gorilla/websocket"
+	"github.com/rs/zerolog"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/suite"
+
+	dpmock "github.com/onflow/flow-go/engine/access/rest/websockets/data_provider/mock"
+	connectionmock "github.com/onflow/flow-go/engine/access/rest/websockets/mock"
+	"github.com/onflow/flow-go/engine/access/rest/websockets/models"
+	"github.com/onflow/flow-go/utils/unittest"
+)
+
+// ControllerSuite is a test suite for the WebSocket Controller.
+type ControllerSuite struct {
+	suite.Suite
+
+	logger zerolog.Logger
+	config Config
+
+	connection          *connectionmock.WebsocketConnection
+	dataProviderFactory *dpmock.DataProviderFactory
+}
+
+func TestControllerSuite(t *testing.T) {
+	suite.Run(t, new(ControllerSuite))
+}
+
+// SetupTest initializes the test suite with required dependencies.
+func (s *ControllerSuite) SetupTest() {
+	s.logger = unittest.Logger()
+	s.config = Config{}
+
+	s.connection = connectionmock.NewWebsocketConnection(s.T())
+	s.dataProviderFactory = dpmock.NewDataProviderFactory(s.T())
+}
+
+// TestConfigureConnection ensures that the WebSocket connection is configured correctly.
+func (s *ControllerSuite) TestConfigureConnection() {
+	controller := s.initializeController()
+
+	// Mock configureConnection to succeed
+	s.mockConnectionSetup()
+
+	// Call configureConnection and check for errors
+	err := controller.configureConnection()
+	s.Require().NoError(err, "configureConnection should not return an error")
+
+	// Assert expectations
+	s.connection.AssertExpectations(s.T())
+}
+
+// TestControllerShutdown ensures that HandleConnection shuts down gracefully when an error occurs.
+func (s *ControllerSuite) TestControllerShutdown() {
+	s.T().Run("keepalive routine failed", func(*testing.T) {
+		controller := s.initializeController()
+
+		// Mock configureConnection to succeed
+		s.mockConnectionSetup()
+
+		// Mock keepalive to return an error
+		done := make(chan struct{}, 1)
+		s.connection.On("WriteControl", websocket.PingMessage, mock.Anything).Return(func(int, time.Time) error {
+			close(done)
+			return websocket.ErrCloseSent
+		}).Once()
+
+		s.connection.
+			On("ReadJSON", mock.Anything).
+			Return(func(interface{}) error {
+				_, ok := <-done
+				if !ok {
+					return websocket.ErrCloseSent
+				}
+				return nil
```

Comment:
it doesn't look like this could return nil. if not, make that clear in the code to avoid confusion by the next person to modify these tests
```suggestion
				<-done
				return websocket.ErrCloseSent
```

Reply:
[2e8f1bc](https://github.com/onflow/flow-go/pull/6757/commits/2e8f1bcc99185c7c4522b015ae8ca5eff648e29c)